### PR TITLE
feat

### PR DESCRIPTION
**Pull request by rohitxsh**
Add functionality to automatically upload parquet files to AWS S3
Add consts
Add typings
Fix and update readme
Add logging
Add all the tables for human and mouse
Add dockerfile

### CODE REVIEW

File: code.py [hunk 1]
```diff
@@ -1,32 +1,46 @@
...
-        Dictionary containing the parsed data.
+        Dictionary containing the parsed data
 
     Raises:
-        FileNotFoundError: If `config_file` doesn't exist.
-    """
+        FileNotFoundError: If `config_file` doesn't exist
+    '''
 
     try:
```
Comment (andrewyatz) [inline code.py:28]: If we can't read the file shouldn't we just fail and kill the process right away? Also you said in the docs this raises a file not found error but we're trapping it here if I am reading this correctly

**Review by andrewyatz — CHANGES_REQUESTED** [review]
Looks good on the whole on my first read through. Note the query about config exception handling

File: code.py (hunk 1 above)
Comment (rohitxsh) [reply code.py:29]: @andrewyatz Updated the PR with a new commit which should handle this, please check and re-review

**Review by andrewyatz — APPROVED** [review]
Changes approved

**Comment by rohitxsh** [comment]
@marcoooo Updated the PR as per review your comments and suggestions, please review

**Review by andrewyatz — APPROVED** [review]
Looks excellent to me

File: sql2parquet/sql2parquet.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+            db_connection_str = f'mysql+pymysql://{database["db_user"]}@{database["location"]}:{database["port"]}/{species["db_name"]}'
+            engine = create_engine(db_connection_str)
+
+            for table in database['tables']:
+                logger.info('Exporting %s table for %s...', table["table_name"], species["species_name"])
+
+                df = pd.read_sql(table['query'], con=engine, params={'species_name': species['species_name']})
+
+                if not os.path.exists(os.path.join('parquet', f'data={table["table_name"]}')): os.mkdir(os.path.join('parquet', f'data={table["table_name"]}'))
```
Comment (bilalebi) [inline sql2parquet/sql2parquet.py:63]: `'parquet'` param can be replaced with `OUTPUT_DIRECTORY`
```
if not os.path.exists(os.path.join(OUTPUT_DIRECTORY, f'data={table["table_name"]}')): os.mkdir(os.path.join(OUTPUT_DIRECTORY, f'data={table["table_name"]}'))
```
Same for the line below (line 65)

File: sql2parquet/sql2parquet.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            # truncate current working directory path to get relative path
+            file = os.path.relpath(file, os.getcwd())
+            # put all the sub directories directly in the S3 bucket, not inside OUTPUT_DIRECTORY folder
+            awsPath = file.replace(os.path.join(OUTPUT_DIRECTORY,''), '')
+            # for Windows machines convert windows path to UNIX path
+            if os.name == 'nt': awsPath = str(Path(awsPath).as_posix())
+
+            try:
+                s3_client.upload_file(file, AWS_S3_BUCKET, awsPath)
+            except ClientError as e:
```
Comment (bilalebi) [inline sql2parquet/sql2parquet.py:98]: The script dies silently when I didn't add the proper AWS credentials, logs show the following error
```
botocore.exceptions.NoCredentialsError: Unable to locate credentials
2022-07-04 14:01:16,053 DEBUG utils s3transfer.utils Releasing acquire 0/None
2022-07-04 14:01:16,053 ERROR sql2parquet __main__ Unable to locate credentials
```
I think it is worth capturing it.

**Review by bilalebi — CHANGES_REQUESTED** [review]
You are doing a great job @rohitxsh, I added some comments to make sure the script runs properly

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,12 @@
+FROM python:3.9
+
+COPY sql2parquet/__init__.py /sql2parquet/__init__.py
```
[the remark this answers is not in the document]
Comment (rohitxsh) [reply Dockerfile:3]: If we are run the script locally even once then there will be files to ignore than to include so kept it as it is, still should I go with .dockerignore instead?

File: sql2parquet/sql2parquet.py (hunk 2 above)
Comment (rohitxsh) [reply sql2parquet/sql2parquet.py]: Thank you for pointing it out, don't know how I missed it, must have messed it up while refactoring
Fixed it in latest commit

File: sql2parquet/sql2parquet.py (hunk 3 above)
Comment (rohitxsh) [reply sql2parquet/sql2parquet.py:98]: Handled NoCredentialsError exception in latest commit

File: sql2parquet/sql2parquet.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
+from botocore.exceptions import ClientError
+from pathlib import Path, PureWindowsPath, PurePosixPath
+from sqlalchemy import create_engine
+from time import time
+from typing import Dict
+import pandas as pd
+import boto3, glob, logging, os, sys, toml
+
+CONFIG_FILE_NAME = 'config.toml'
```
[the remark this answers is not in the document]
Comment (bilalebi) [reply sql2parquet/sql2parquet.py:9]: I see, that makes sense, my bad :)

**Review by bilalebi — APPROVED** [review]

File: sql2parquet/sql2parquet.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        logger.error(e)
+        sys.exit(e)
+
+    logger.info('Uploading the parquet files to AWS S3...')
+    try:
+        uploadDirToS3()
+    except Exception as e:
+        logger.error(e)
+
+    logger.info('Done, it took a total of %s sec(s).', (time()-start))
```
[the remark this answers is not in the document]
Comment (rohitxsh) [reply sql2parquet/sql2parquet.py]: Handled in the latest commit

File: config.toml [hunk 7]
```diff
@@ -21,4 +24,152 @@ JOIN transcript ct on (g.canonical_transcript_id = ct.transcript_id)
...
+JOIN object_xref ox on (tr.translation_id = ox.ensembl_id and ox.ensembl_object_type = 'Translation')
+JOIN xref x on (ox.xref_id = x.xref_id)
+JOIN external_db edb ON (x.external_db_id = edb.external_db_id)
+JOIN seq_region sr ON (g.seq_region_id = sr.seq_region_id)
+JOIN coord_system cs ON (sr.coord_system_id = cs.coord_system_id)
+JOIN meta m on (cs.species_id = m.species_id and m.meta_key = "species.production_name")
+WHERE g.biotype <> "LRG_gene"
+AND m.meta_value = 'homo_sapiens'
+AND m.meta_value = %(species_name)s
 '''
```
[the remark this answers is not in the document]
Comment (rohitxsh) [reply config.toml]: Handled in the latest commit

**Review by marcoooo — APPROVED** [review]

**Pull request merged**